pkg/gds/admin.go

```diff
@@ -149,22 +160,224 @@ func (s *Admin) setupRoutes() (err error) {
+		return
+	}
+
+	// Validate the credential with Google
+	if claims, err = idtoken.Validate(c.Request.Context(), in.Credential, s.conf.Audience); err != nil {
+		log.Warn().Err(err).Msg("invalid credentials used for authentication")
+		c.JSON(http.StatusUnauthorized, admin.ErrorResponse("invalid credentials"))
+		return
+	}
+
+	// Verify that the domain is one of our authorized domains
+	if err = s.checkAuthorizedDomain(claims); err != nil {
+		log.Warn().Err(err).Msg("access request from unauthorized domain")
+		c.JSON(http.StatusUnauthorized, admin.ErrorResponse("invalid credentials"))
+		return
+	}
+
+	// Create the access and refresh tokens from the claims
+	if accessToken, err = s.tokens.CreateAccessToken(claims); err != nil {
+		log.Error().Err(err).Msg("could not create access token")
+		c.JSON(http.StatusInternalServerError, admin.ErrorResponse("could not authenticate with credentials"))
+		return
+	}
+
+	if refreshToken, err = s.tokens.CreateRefreshToken(accessToken); err != nil {
+		log.Error().Err(err).Msg("could not create refresh token")
+		c.JSON(http.StatusInternalServerError, admin.ErrorResponse("could not authenticate with credentials"))
+		return
+	}
+
+	// Sign the tokens and return the response
+	out = new(admin.AuthReply)
+	if out.AccessToken, err = s.tokens.Sign(accessToken); err != nil {
+		log.Error().Err(err).Msg("could not sign access token")
+		c.JSON(http.StatusInternalServerError, admin.ErrorResponse("could not authenticate with credentials"))
+		return
+	}
+	if out.RefreshToken, err = s.tokens.Sign(refreshToken); err != nil {
+		log.Error().Err(err).Msg("could not sign refresh token")
+		c.JSON(http.StatusInternalServerError, admin.ErrorResponse("could not authenticate with credentials"))
+		return
+	}
+
+	// Refresh the double cookies for CSRF protection while using the access/refresh tokens
+	expiresAt := refreshToken.Claims.(*tokens.Claims).ExpiresAt
+	if err := admin.SetDoubleCookieTokens(c, expiresAt); err != nil {
 		c.JSON(http.StatusInternalServerError, admin.ErrorResponse("could not set cookies"))
 		return
 	}
 
-	c.JSON(http.StatusNotImplemented, admin.ErrorResponse("not implemented yet"))
+	// Return successful authentication!
+	c.JSON(http.StatusOK, out)
 }
 
+func (s *Admin) checkAuthorizedDomain(claims *idtoken.Payload) error {
+	// Fetch the claim from the idtoken payload
+	domain, ok := claims.Claims["hd"]
+	if !ok {
+		return errors.New("no hd claim to verify authorized domain with")
+	}
+
+	// Convert the domain into a string for verification
+	domains, ok := domain.(string)
+	if !ok {
+		return fmt.Errorf("claim type %T unparseable", domain)
+	}
+
+	// Search the authorized domains, if found return nil
+	for _, authorized := range s.conf.AuthorizedDomains {
+		if domains == authorized {
+			// Found an authorized domain!
+			return nil
+		}
+	}
+
+	return fmt.Errorf("%s is not in the configured authorized domains", domains)
```

Comment:
Is the idea that if this error comes up, the UI should consume it and prompt the user to log in with a different google account?

Reply:
Yes, e.g. if you accidentally signed in with your @gmail.com account.